### ISSUE TITLE
Observable effects completion statuses for better server rendering

### ISSUE THREAD

**Issue by bardt** [redux-loop/redux-loop#23]
I was in the middle of brainstorm about better server-rendering and side effects management in redux apps when I found this repo. I want to share some thoughts here and maybe take part in developing if you find them sensible.

## tl;dr

Having a way to observe the effects completion statuses (like how many effects are not completed yet) could give us an ability to completely separate UI and application state. By achieving this we could get much simpler server-rendering techniques.

## Current server-rendering approach

I assume React as a view layer below, but conclusions are framework-agnostic.

The main difference in server and client rendering is the need to wait for all the API calls and other side effects finish before first rendering to string. These days this is solved by specifying which effects a UI component should perform (ex. API calls to be performed), getting the list of all the effects according to a route, perform all of them, dispatch resulting actions through redux store and then render it all only once. A good example of this approach is  [redial](https://github.com/markdalgleish/redial) server-rendering documentation.

Managing side effects differently during server rendering and client initialization feels wrong for me. I tend to believe there should be a solution to write effects management once and reuse it on both server and client sides.

## Main idea

Making decisions on next effects as close to the store as possible (which is done thanks to this project) gives us an advantage, because we separate all possible state transitions from UI.

Let's assume we have a queue of effects. When a reducer returns an effect, we put it into a queue. When this effect is resolved, it removes itself from queue. So, looking on the queue length we can definitely say if our system is stable (came to the state there we no longer wait for anything despite user input).

Given this ability, server-rendering is simplified to these steps:
1. Get all the initial parameters from router.
2. Init store with these parameters.
3. Subscribe to store changes. On each store change check if effects queue is empty.
4. Once the queue is empty we assume system is stable and there is nothing to wait for. So, we perform server rendering. Done.

Client initialization has no difference, except we don't have to check queue emptiness.

## Potential problems

I believe there could be an effect that is not ever resolved in classical meaning; getting data from a websocket or infinite timers are examples which come to my mind. For such effects, there should probably be a `infinite` flag to avoid taking them into account.

**Comment by lukewestby** [comment]
Hey @bardt, thanks so much for contributing your ideas. I'm not feeling super well this weekend and want to give this the full attention it deserves so I'll get back to you in a couple days.

**Comment by lukewestby** [comment]
So! @bardt, thank you again for contributing your thoughts.

Adding this as a distinct feature shouldn't be too difficult, the basis for this functionality is already available. That is, when you dispatch an action dispatch returns a Promise for when that action and any future actions it causes have been completed. We can use this to set up a subscription-based mechanism for listening for the completion of an effect path.

Regarding infinite actions, redux-loop is geared toward first-order reactive programming only, meaning that using a websocket or other long-lived data sources would look more like an `input` in Elm's `start-app`. You would initialize the websocket outside your store and then dispatch actions on the store directly from within the subscription to that socket. In short, we don't need to worry about it! :smile: 

What do you imagine the API for this would look like? Once we have an idea of what you'd like to see I can provide any initial guidance and then you're free to start hacking and submit a PR!

**Comment by bardt** [comment]
@lukewestby from the very top level, I'd love to extend store with new entity – `effectsQueue`. The code could look like this:

```
store.effectsQueue; // [ [objectObject], [objectObject] ]
store.effectsQueue.isEmpty(); // false

// For the subscription
store.subscribe(() => {
  if (store.effectsQueue.isEmpty()) {
    runServerRendering();
  } else {
    // Do nothing, just wait for state to stabilise
  }
}); 
```

This means, the queue should be updated before subscription listeners are fired. I haven't read your code attentively yet (sorry for that), so I'm not sure if it fits the whole architecture. Looks like we should overwrite subscribe method to be fired a bit later, because now effects are processed [after listeners are fired](https://github.com/raisemarketplace/redux-loop/blob/master/modules/install.js#L44).

I'd start from modifying `runEffect` [function](https://github.com/raisemarketplace/redux-loop/blob/master/modules/install.js#L49) to manage adding/removing effects from queue.

Naming is also a question :)

**Comment by lukewestby** [comment]
@bardt something to consider:

For this case are you only interested in whether the queue is empty or do you want to inspect the items in the queue? In the case that you only want to be notified when the queue has emptied I think a callback-based subscription similar to `subscribe` might work better.

**Comment by bardt** [comment]
@lukewestby for me it is strange to provide `isEmpty` as callback and not give ability to inspect the queue. I don't see any reasons why anyone should want to inspect the queue items, but that doesn't mean there aren't such cases. The more flexible solution is preferred.

We could provide a subscribe solution as well as manual check, 'cause I think these things are related under the hood.

Do you have any suggestions on naming? Do I need to know anything important before I start hacking?

**Comment by lukewestby** [comment]
@bardt I disagree with your point about including a more flexible solution in this case, but please don't take this to mean that any future suggestions you have aren't valued!

I don't want to include functionality without a demonstrable reason for why it's needed. Since you've said yourself that you don't see a reason why someone would want access to the active effects in the queue then let's leave it out of this feature and if a real use cases become apparent we'll add it later. Since the concepts this library introduces are frequently very new to JavaScript developers with no Elm experience I don't want to add to the learning curve with features where we can't say with total precision what their purpose is.

Regarding callback vs. manual check, though, I'm totally onboard with manually checking if the queue is empty if you think a callback-based approach wouldn't work as well. What do you think about a getter which mirrors your original suggestion, like:

```js
store.subscribe(() => {
  if (store.isEffectsQueueEmpty()) {
    runServerRendering();
  } else {
    // Do nothing, just wait for state to stabilise
  }
}); 
```

so we have the manual check without introducing as-of-yet unneeded functionality?

**Comment by bardt** [comment]
@lukewestby I totally understand your position about flexible solution here. That's not a point for discussion right now, I agree to add features as soon as real cases appear. 

That way, lets check pros and cons of a separate queue subscription mechanics. 

Pros:
+ No need to modify store subscription internally
Cons:
- Not obvious if it is fired once or multiple times during lifecycle
- Difficult to check queue emptiness later (ex. inside middlewares)

Did I forget anything? Currently I tend to prefer manual check.

**Comment by lukewestby** [comment]
@bardt looks like a manual check is the right way to go to me!

**Comment by bardt** [comment]
@lukewestby I will start hacking during next week then. Do you want to give me any instructions?

**Comment by lukewestby** [comment]
@bardt I think we've covered everything! I haven't gotten around to adding an eslintrc yet so just try to stick to the existing coding style as much as you can. It'll be on me to make sure it's totally consistent later. Let me know if you have any questions along the way. Thanks!

**Comment by bardt** [comment]
@lukewestby check my PR, please. The simplest solution I came to is there. Another question: I modified the example to check my general logic. Should I create a separate example to demonstrate server-side rendering or modify existing one to use a simple server?

**Comment by lukewestby** [comment]
@bardt sorry about the delay, I'll give it a thorough review on Monday. Regarding an example, if you have time to create a separate example demonstrating how you want to use this for server-side rendering that would be fantastic. If not, your addition to the current example seems plenty sufficient to communicate the usage of the feature.

**Comment by buzinas** [comment]
Any news on this? @bardt @lukewestby

**Comment by bardt** [comment]
@buzinas I've kind of switched to another task for some time. Plan to work on some examples in this PR tomorrow.
